src/Common/src/System/Text/DBCSDecoder.cs

```diff
@@ -0,0 +1,293 @@
+            if (b < _leadyByteRanges[0])
+                return false;
+            int i = 0;
+            while (i < _rangesCount)
+            {
+                if (b >= _leadyByteRanges[i] && b <= _leadyByteRanges[i + 1])
+                    return true;
+                i += 2;
+            }
+            return false;
+        }
+
+        public override void Reset()
+        {
+            _leftOverLeadByte = 0;
+        }
+
+        public override unsafe int GetCharCount(byte[] bytes, int index, int count)
+        {
+            return GetCharCount(bytes, index, count, false);
+        }
+
+        public override unsafe int GetCharCount(byte[] bytes, int index, int count, bool flush)
+        {
+            if (bytes == null)
+                throw new ArgumentNullException(nameof(bytes), SR.ArgumentNull_Array);
+
+            if (index < 0 || count < 0)
+                throw new ArgumentOutOfRangeException(index < 0 ? nameof(index) : nameof(count), SR.ArgumentOutOfRange_NeedNonNegNum);
+
+            if (bytes.Length - index < count)
+                throw new ArgumentOutOfRangeException(nameof(bytes), SR.ArgumentOutOfRange_IndexCountBuffer);
+            
+            if (count == 0 && (_leftOverLeadByte == 0 || !flush))
+                return 0;
+
+            fixed (byte* pBytes = bytes)
+            {
+                byte dummyByte;
+                byte* pBuffer = pBytes == null ? &dummyByte : pBytes + index;
+                
+                return GetCharCount(pBuffer, count, flush);
+            }
+        }
+
+        private unsafe int ConvertWithLeftOverByte(byte* bytes, int count, char* chars, int charCount)
+        {
+            Debug.Assert(_leftOverLeadByte != 0);
+            byte* pTempBuffer = stackalloc byte[2];
+            pTempBuffer[0] = _leftOverLeadByte;
+
+            int index = 0;
+
+            if (count > 0)
+            {
+                pTempBuffer[1] = bytes[0];
+                index++;
+            }
+
+            int result = OSEncoding.MultiByteToWideChar(_encoding.CodePage, pTempBuffer, index+1, chars, charCount);
+
+            if (count - index > 0)
+                result += OSEncoding.MultiByteToWideChar(
+                                        _encoding.CodePage, bytes + index, 
+                                        count - index,
+                                        chars == null ? null : chars + result, 
+                                        chars == null ? 0 : charCount - result);
+
+            return result;
+        }
+
+        public unsafe override int GetCharCount(byte* bytes, int count, bool flush)
+        {
+            if (bytes == null)
+                throw new ArgumentNullException(nameof(bytes), SR.ArgumentNull_Array);
+
+            if (count < 0)
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedNonNegNum);
+
+            bool excludeLastByte = count > 0 || (!flush && IsLastByteALeadByte(bytes, count));
```

Comment:
If count == 0 && flush is false, we end up calling IsLasByteALeadByte with bytes and 0, and then IsLastByteALeadByte ends up accessing bytes[-1].  Is that ok?